src/devices/src/virtio/queue.rs

```diff
@@ -165,6 +168,60 @@ pub struct Queue {
     next_used: Wrapping<u16>,
 }
 
+#[derive(Clone, Debug, Versionize)]
+pub struct QueueState {
+    /// The maximal size in elements offered by the device
+    max_size: u16,
+
+    /// The queue size in elements the driver selected
+    pub size: u16,
+
+    /// Indicates if the queue is finished with configuration
+    pub ready: bool,
+
+    /// Guest physical address of the descriptor table
+    pub desc_table: u64,
+
+    /// Guest physical address of the available ring
+    pub avail_ring: u64,
+
+    /// Guest physical address of the used ring
+    pub used_ring: u64,
+
+    next_avail: Wrapping<u16>,
+    next_used: Wrapping<u16>,
+}
+
+impl Into<QueueState> for Queue {
+    fn into(self) -> QueueState {
+        QueueState {
+            max_size: self.max_size,
+            size: self.size,
+            ready: self.ready,
+            desc_table: self.desc_table.0,
+            avail_ring: self.avail_ring.0,
+            used_ring: self.used_ring.0,
+            next_avail: self.next_avail,
+            next_used: self.next_used,
+        }
+    }
+}
+
+impl From<QueueState> for Queue {
+    fn from(state: QueueState) -> Self {
+        Queue {
+            max_size: state.max_size,
+            size: state.size,
+            ready: state.ready,
+            desc_table: GuestAddress::new(state.desc_table),
+            avail_ring: GuestAddress::new(state.avail_ring),
+            used_ring: GuestAddress::new(state.used_ring),
+            next_avail: state.next_avail,
+            next_used: state.next_used,
+        }
+    }
+}
+
```

Comment:
This is a non-negligible amount of hand-written code to avoid deriving `Versionize` on `GuestAddress`.

If this is the case in many places around the code-base I am inclined to just `Versionize` `GuestAddress` (at least for now - we don't know yet what the final solution in rust-vmm will be).

Reply:
I was not very explicit to why I think there should be a `QueueState`. I will try to incorporate this in the PR somehow.

Just a few thoughts:

Not implementing serialize for `GuestAddress` is just one of the reason for keeping a `QueueState` and it might not be a strong one. What I find to be a strong reason is that I would like the serialization to be completely separated from the device logic. For that matter, I think all the serialization related things (that can be moved), should be in a separate module called `snapshot.rs` or  something to that extend. The reason is that we are a bit behind in rust-mvm on serialization, and we didn't yet agree on what will be the format. We also didn't decided if we are going to have cross-version support. If we strongly tie the implementation of devices (which we know we want to consume from rust-vmm) with the serialization, than there is a higher risk and we might need to make this separation anyway. Since we didn't yet serialized the devices, I think it is a good opportunity to find a way by which future changes (i.e. when consuming devices from rust-vmm) are unlikely to to happen or do not affect a large portion of the code.

When looking at this code in particular, I don't have a huge concern because there are only 50 lines of code and they are going to be used by all the devices.

If you're also okay with this, I would say to discuss this at length on Monday offline.

Reply:
If we look at `GuestAddress` as a primitive, then `Queue` becomes its own state, and we can simply use a type alias to define `QueueState` as `Queue`. We might just do that here for now, since `Queue` is one of the basic building blocks we plan to get entirely from rust-vmm anyway when it's ready (as opposed to a particular device, which might end up as custom glue on top of rust-vmm building blocks).

At the same time, we can also keep the code added here since it's quite self contained, but at the very least we have to decide how we think `QueueState` will end up looking (i.e. `GuestAddress` vs `u64`).